### ISSUE TITLE
Convert between => and { return ...; } based on line length (WONTFIX)

### ISSUE THREAD

**Issue by Hixie** [dart-lang/dart_style#452]
It would be interesting if the formatter could automatically add or remove braces based on the line length.

For example, turning this:

``` dart
   bool get hasSize {
     return true;
   }
   int crazyLongFunction(int withMany, int manyMany, int arguments) => withMany + manyMany + arguments +
      withMany * 2 + manyMany * 5 + arguments;
```

...into this:

``` dart
   bool get hasSize => true;
   int crazyLongFunction(int withMany, int manyMany, int arguments) {
     return withMany + manyMany + arguments + withMany * 2 + manyMany * 5 + arguments;
   }
```

**Comment by munificent** [comment]
Like I mentioned on #451, this is out of scope for the formatter as it's currently specced.

**Comment by Hixie** [comment]
In that case, I think there should be a restriction that the formatter never insert a line break into an expression that contains a `=>` operator.

**Comment by jmesserly** [comment]
I think it's time to move past the "only whitespace" rule. It's a self-imposed restriction. The theory was to help adoption by making the formatter seem less scary. But it seems to be doing the opposite, by preventing improvements that users would like to see.

**Comment by munificent** [comment]
See the comments on #451. I think it's a useful restriction. Making more significant changes is a *lot* more complex and has a lot more failure modes. I don't think that's a good default behavior for the formatter. That's what the linter package is for.

**Comment by alan-knight** [comment]
Also, typical style for Intl messages is as lambda's, even though they might be several lines. I wouldn't like having 

      hello(name) => Intl.message("Hello $name", 
          name: "hello", 
          desc: "Say Hello to the user",
          examples: { "name" : "Alice"});

turned into a function with braces because it didn't fit.

**Comment by Hixie** [comment]
```dart
String hello(name) {
  return Intl.message(
    'Hello $name',
    name: 'hello',
    desc: 'Say Hello to the user',
    examples: <String, String>{ 'name': 'Alice' }
  );
}
```

**Comment by kasperpeulen** [comment]
I actually like how dart fmt handles this now:
![image](https://cloud.githubusercontent.com/assets/1035299/11102170/4465e6d0-88bb-11e5-8c81-a03ce1e81654.png)

Adding the return statement, would make the line a bit longer then 80 chars:
![image](https://cloud.githubusercontent.com/assets/1035299/11102200/68dbf2c0-88bb-11e5-8eb5-a92d01566cd8.png)

which would give this as result:
![image](https://cloud.githubusercontent.com/assets/1035299/11102211/7761e516-88bb-11e5-8158-6b94c584f4b3.png)

I'm not totally convinced if dart fmt should only handle whitespace, but in this case, I'm glad that it doesn't convert my `=>` in `{ return ...; }`.

**Comment by Hixie** [comment]
I agree that the version with the wrapped return looks bad, but that's because of the misaligned "withMany" vs "manyMany", IMHO. It still lookes better than the newline after the `=>`.

Personally I am in favour of "soft" 80 character limits which would lead to the second one above.

But the ideal wrapping for that particular function is probably the following:
```dart
int crazyLongFunction(int withMany, int manyMany, int arguments) {
  return withMany     + manyMany     + arguments +
         withMany * 2 + manyMany * 5 + arguments;
}
```
None of the other wrappings should what's actually going on as well as that one. And once you see that, you quickly see that the code should really be:
```dart
int crazyLongFunction(int withMany, int manyMany, int arguments) {
  return withMany * 3 + manyMany * 6 + arguments * 2;
}
```

**Comment by kasperpeulen** [comment]
@Hixie I agree about the symmetry argument (you also raised somewhere else I think). It would be nice if dartfmt could do something like:

```dart
int crazyLongFunction(int withMany, int manyMany, int arguments) {
  return withMany +
         manyMany +
         arguments +
         withMany * 2 +
         manyMany * 5 +
         arguments;
}
```

I don't really see though why this: 
```dart
int crazyLongFunction(int withMany, int manyMany, int arguments) {
  return withMany * 3 + manyMany * 6 + arguments * 2;
}
```

would be prefered over:
```dart
int crazyLongFunction(int withMany, int manyMany, int arguments) =>
    withMany * 3 + manyMany * 6 + arguments * 2;
```

I always format it with an arrow, where possible.

**Comment by Hixie** [comment]
At the end of the day, it's a matter of personal preference. I find that `=>` works well when you have a very short expression, because then it looks like you're doing something cheap. I find that if you're doing something long, then `{ ... }` looks better because it's more obvious that it's a bunch of code.

If I am scanning through code and see the line:

```withMany * 3 + manyMany * 6 + arguments * 2;```

...then it's not obvious to me that it's part of a function unless it's wrapped in braces. The braces are aligned so that the closing brace lines up with the start of the opening lines, so I can easily tell which block the code is in. When there's a bunch of lines:

```dart
int crazyLongFunction(int withMany, int manyMany, int arguments) =>
    withMany * 3 + manyMany * 6 + arguments * 2;
int crazyLongFunction(int withMany, int manyMany, int arguments) =>
    withMany * 3 + manyMany * 6 + arguments * 2;
int crazyLongFunction(int withMany, int manyMany, int arguments) =>
    withMany * 3 + manyMany * 6 + arguments * 2;
```

...then I find it much harder to tell what's going on, because to interpret the indented lines, I have to look at the end of the previous line. With blocks:

```dart
int crazyLongFunction(int withMany, int manyMany, int arguments) {
    return withMany * 3 + manyMany * 6 + arguments * 2;
}
int crazyLongFunction(int withMany, int manyMany, int arguments) {
    return withMany * 3 + manyMany * 6 + arguments * 2;
}
int crazyLongFunction(int withMany, int manyMany, int arguments) {
    return withMany * 3 + manyMany * 6 + arguments * 2;
}
```
...then I don't have to move my eyeline away from the left margin. I can immediately tell that the expression is in a block, I can immediately tell it's a return value, I can immediately tell where the block begins.

**Comment by Hixie** [comment]
Now that `dartfmt` has the `--fix` mode, it seems like this is now in-scope again, FWIW.

**Comment by munificent** [comment]
Yes, this could be a fix.

In practice, I'm not sure how many users would actually apply it. I see many member bodies that use `=>` and where the expression is longer than one line.

Technically speaking, this is challenging to implement because of how the formatter is architected. It doesn't know how long a line is until after calculating line splitting and indentation. But once it's done that, it doesn't expect the text to change. For a fix, though, we might be able to do something clever/hacky/slow (like format it twice) since fixes aren't as perf critical as general formatting.

**Comment by leecommamichael** [comment]
I disliked having this fix enabled in Javascript. It's not fun fiddling an arrow-function into a block-scoped function. It would be clunkier in Dart, since `=>` is only valid for lambdas.

This was usually to sneak a `console.log` somewhere that had been tightly crafted, but it still annoyed me. As ever, it's the tradeoff between dead-simple and beautiful.

**Comment by munificent** [comment]
I'm going to close this since we're getting rid of `dart format --fix` in favor of `dart fix`: #1153. We want all non-whitespace code changes to be handled by the much more powerful `dart fix`. I believe it may already have lint rules and quick fixes around the different forms of function bodies.
